File: 20221220_Tuesday/20221220.go
// Leetcode: 841. Keys and Rooms.
// https://leetcode.com/problems/keys-and-rooms/
// = = = = = = = = = = = = = =
// Accepted.
// Thanks God, Jesus Christ!
// = = = = = = = = = = = = = =
// Runtime: 12 ms, faster than 37.10% of Go online submissions for Keys and Rooms.
// Memory Usage: 4.1 MB, less than 83.87% of Go online submissions for Keys and Rooms.
// 2022.12.20 Daily Challenge.
func canVisitAllRooms(rooms [][]int) bool {
  n := len(rooms)
  counter := 0
  v := make([]bool,n)
  q := make([]int,0)
  q = append(q,0)
  v[0] = true
  for 0 != len(q) {
    x := q[0]
    q = q[1:len(q)]
    counter += 1
    for _,y := range rooms[x] {
      if v[y] {
        continue
      }
      v[y] = true
      q = append(q,y)
    }
  }
  return n == counter
}

File: 20221220_Tuesday/20221220.rb
# Leetcode: 841. Keys and Rooms.
# https://leetcode.com/problems/keys-and-rooms/
# = = = = = = = = = = = = = =
# Accepted.
# Thank God, Jesus Christ!
# = = = = = = = = = = = = = =
# Runtime: 163 ms, faster than 28.57% of Ruby online submissions for Keys and Rooms.
# Memory Usage: 211.2 MB, less than 85.71% of Ruby online submissions for Keys and Rooms.
# 2022.12.20 Daily Challenge.
# @param {Integer[][]} rooms
# @return {Boolean}
def can_visit_all_rooms(rooms)
  n = rooms.size
  counter = 0
  v = Array.new(n,false)
  q = [0]
  v[0] = true
  while ! q.empty?
    x = q.shift()
    counter += 1
    rooms[x].each do |y|
      next if v[y]
      v[y] = true
      q.push(y)
    end
  end
  counter == n
end
